### PR TITLE
basic sub nested insertion

### PR DESCRIPTION
**Pull request by olafbuitelaar**
Since commit; https://github.com/ClickHouse/ClickHouse/pull/17310 clickhouse supports nested sub-columns.
I'm trying to implement this in this library. It still needs some work for many level's. But insertion of 1 level sub-columns works.
Clickhouse seems to tread them as arrays of tuple's. So the parse add's a `virtual` array around the nested construct.
Only somehow i don't seem to get the BinaryStreamReader to add the tuple's to the array. The tuple it self is read fine, but it fails in adding it to the array with an invalid cast exception, while the types appear to be the same. Would you have any clue?
=> see comment, NestedType, always did an extra conversion to Array, causing a double Array type, which produced the invalid cast error

### CODE REVIEW

File: ClickHouse.Client/Formats/BinaryStreamReader.cs [hunk 1]
```diff
@@ -120,7 +131,19 @@ public object Read(IPv4Type pv4Type)
...
+            // => throw new NotSupportedException();
+            var count = tupleType.UnderlyingTypes.Length;
+            var contents = new object[count];
+            for (var i = 0; i < count; i++)
+            {
+                var value = Read(tupleType.UnderlyingTypes[i]);
+                contents[i] = ClearDBNull(value);
+            }
+            //return (1, "vx1" );
+            return tupleType.MakeTuple(contents);
```
Comment (olafbuitelaar) [inline ClickHouse.Client/Formats/BinaryStreamReader.cs:145]: Tuple is read fine here

File: ClickHouse.Client/Formats/BinaryStreamReader.cs [hunk 2]
```diff
@@ -44,7 +47,15 @@ public object Read(ArrayType arrayType)
             var data = arrayType.MakeArray(length);
             for (var i = 0; i < length; i++)
             {
-                data.SetValue(ClearDBNull(Read(arrayType.UnderlyingType)), i);
+                var val = ClearDBNull(Read(arrayType.UnderlyingType));
+                if (arrayType.UnderlyingType.TypeCode == ClickHouseTypeCode.Nested)
+                {
+                    //data.SetValue(val, i);
```
Comment (olafbuitelaar) [inline ClickHouse.Client/Formats/BinaryStreamReader.cs:53]: Insertion fails here

Comment (olafbuitelaar) [reply ClickHouse.Client/Formats/BinaryStreamReader.cs]: The NestedType appeared to have overwritten the propery FrameworkType => base.FrameworkType.MakeArrayType();
causing a double array. I'm not sure why this is, but added a check if the type is already of an Array, and only when not convert it to an Array type.
All unit tests pass now.

**Comment by olafbuitelaar** [comment]
The failing tests seems because they are ran against clickhouse version: 20.3.18.10, i've tested with version; 21.5.5.12

**Review by DarkWanderer — CHANGES_REQUESTED** [review]
* Grammar parser must not know about types
* If something fails in regression test, it means that it will equally fail in produciton when using older version. Please use feature flags to gate the new functionality

File: ClickHouse.Client/Types/Grammar/Parser.cs [hunk 3]
```diff
@@ -8,25 +8,60 @@ public static class Parser
...
+                        level--;
+                        if (current.Virtual)
+                        {
+                            current = stack.Pop();
+                            level--;
+                        }
                         break;
                     default:
-                        current = new SyntaxTreeNode { Value = token };
+                        if ((token == "Nested" || token.EndsWith(" Nested")) && level >= 1)
```
[the remark this answers is not in the document]
Comment (olafbuitelaar) [reply ClickHouse.Client/Types/Grammar/Parser.cs:43]: Clickhouse returns nested types in different ways depending on the level it's defined;
Level 0: Array(<type>) Ex; Array(String)
Level 1: Array(Nested(<fields>) Ex; Array(Nested(param_id UInt8, param_val String))
Level 2: <field name> Nested(<fields>) Ex; Array(Nested(param_id UInt8, param_val String, **sub_sub_nested Nested(param_id UInt8, param_val String)**))

I wouldn't say the parser knows anything about the types, but the way the string tokens can occur for a specific type.
Probably if you don't want to have any knowledge about the string to parse, this needs to be refactored to full-blown AST, to cope with the difference of definition. This seems to be a bit overkill.

What do you propose?

[the remark this answers is not in the document]
Comment (olafbuitelaar) [reply ClickHouse.Client/Types/Grammar/Parser.cs:43]: What feature flag would you propose? This functionality also should only fail when you actually use nested sub-columns in the data set, which you wouldn't if you don't have a compatible clickhouse server.

[the remark this answers is not in the document]
Comment (DarkWanderer) [reply ClickHouse.Client/Types/Grammar/Parser.cs:43]: > I wouldn't say the parser knows anything about the types, but the way the string tokens can occur for a specific type.

You have specific behavior there for specific type which does not occur for other types

> Probably if you don't want to have any knowledge about the string to parse, this needs to be refactored to full-blown AST, to cope with the difference of definition. This seems to be a bit overkill.

Not sure why it needs to be "refactored into full-blown AST"? Right now, parsing `Nested(Nested)` works perfectly:
```
TypeConverter.ParseClickHouseType("Nested (UInt8, String, Nested (UInt8, String,  Nested(UInt8, String)))")
```

If the parser works for this, it works. If the problem is with serialization, it needs to be fixed in serialization.

There are 3 things at play here: 1.Parsing type string into tokens 2.creating a type out of these tokens 3.implementing some behavior based on type. If there is a type `Array(Nested(<fields>))` coming from ClickHouse, it must be represented as types `Array->Nested->[fields]` internally. Right now, you're:

- breaking roundtrip parsing/serialization: `TypeConverter.ParseClickHouseType(type).ToString() != type`
- losing information about parsed type

So basically, this PR introduces a breakage in a fundamental part of code to solve some very specific problem in another part of code

If you need some specific behavior when reading and writing Nested(Nested(...)) values, you need to change BinaryStreamReader/BinaryStreamWriter classes to achieve that

> What feature flag would you propose? This functionality also should only fail when you actually use nested sub-columns in the data set, which you wouldn't if you don't have a compatible clickhouse server.

See new `SupportedFeatures` property in `master` - previously it was a separate set of methods `Supports***()`. If we have different behavior based on server version, it's better to reflect it somewhere in API

File: ClickHouse.Client/Formats/BinaryStreamReader.cs (hunk 1 above)
Comment (DarkWanderer) [reply ClickHouse.Client/Formats/BinaryStreamReader.cs:136]: Shouldn't this return an array of `Tuple`s?

File: ClickHouse.Client/Formats/BinaryStreamWriter.cs [hunk 4]
```diff
@@ -197,7 +197,17 @@ public void Write(EnumType enumType, object value)
             writer.Write(enumIndex);
         }
 
-        public void Write(NestedType nestedType, object value) => throw new NotSupportedException("Writing Nested values directly is not supported, see documentation");
+        public void Write(NestedType nestedType, object value) 
+        {
+            // => throw new NotSupportedException("Writing Nested values directly is not supported, see documentation");
+
+            var items = (object[])value;
```
Comment (DarkWanderer) [inline ClickHouse.Client/Formats/BinaryStreamWriter.cs:204]: I would suggest converting to `IEnumerable` here & using `foreach` - should be able to use `List<T>` just as well

File: ClickHouse.Client/Types/Grammar/Parser.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (olafbuitelaar) [reply ClickHouse.Client/Types/Grammar/Parser.cs:43]: > Not sure why it needs to be "refactored into full-blown AST"? Right now, parsing `Nested(Nested)` works perfectly:

Maybe the parsing works, but CH will thread this as Nested->Array->Tuple. So it add's an additional "virtual" array.

If i understand you correctly you prefer to have separate implementations of the Nested type in the binary reader/writer and evaluate this logic for every row/column in the data-set. Instead of feeding the binary ready/writer with the correct type structure and being only processed once up-front?

I've added the supported feature from your PR #80

File: ClickHouse.Client/Formats/BinaryStreamWriter.cs (hunk 4 above)
Comment (olafbuitelaar) [reply ClickHouse.Client/Formats/BinaryStreamWriter.cs]: Why would you convert the data type here? It's essentially the same as the Tuple implementation.

File: ClickHouse.Client/Formats/BinaryStreamReader.cs (hunk 1 above)
Comment (olafbuitelaar) [reply ClickHouse.Client/Formats/BinaryStreamReader.cs:136]: The value is represented as a single `Tuple`

**Comment by olafbuitelaar** [comment]
It also appears with the PR#80 the test fail related to the uuid in combination with the feature flag `SupportsUUIDParameters`, see; https://github.com/olafbuitelaar/ClickHouse.Client/actions/runs/1033593647

File: ClickHouse.Client/Types/Grammar/Parser.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DarkWanderer) [reply ClickHouse.Client/Types/Grammar/Parser.cs:43]: You are presuming that parsing types as-is is incorrect, but it is not. **Correct** type structure is the one which is on the label, not the one which is "internally". `DateTime64` is internally a `decimal`, `DateTime` is internally a `UInt64`, `FixedString(N)` is internally an array of `char`s. Would you think it's correct to return `decimal` when asked for a `DateTime`?

File: ClickHouse.Client/Formats/BinaryStreamReader.cs (hunk 1 above)
Comment (DarkWanderer) [reply ClickHouse.Client/Formats/BinaryStreamReader.cs:136]: `Nested` is an [array of tuples](https://clickhouse.tech/docs/en/sql-reference/data-types/nested-data-structures/nested/), not a single tuple - how does this work when there are multiple values inside?

File: ClickHouse.Client/Formats/BinaryStreamWriter.cs (hunk 4 above)
Comment (DarkWanderer) [reply ClickHouse.Client/Formats/BinaryStreamWriter.cs]: Because you are already converting to `object[]`, and it significantly narrows down the types you can pass here. `IEnumerable` would allow to use both `List` and array, for example

File: ClickHouse.Client/Types/Grammar/Parser.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (olafbuitelaar) [reply ClickHouse.Client/Types/Grammar/Parser.cs:43]: Of course a `DateTime64` type should be returned that way. I think the misconception is that the data you actually inputting is in the correct format/type, see for example my test case; `new[] { new[] { new object[] { 4, "v4" }, new object[] { 5, "v5" }, new object[] { 6, "v6" } }, new[] { new object[] { 9, "v9" } } }`
While this type structure is not reflected by the type string as returned by CH.

You could also simply run my unittests against your master branch, or simply disable these to;
https://github.com/olafbuitelaar/ClickHouse.Client/blob/master/ClickHouse.Client/Types/Grammar/Parser.cs#L43 
 if ((token == "Nested" || token.EndsWith(" Nested")) && level >= 1) =>  if (**false &&** (token == "Nested" || token.EndsWith(" Nested")) && level >= 1)
and 
https://github.com/olafbuitelaar/ClickHouse.Client/blob/master/ClickHouse.Client/Types/NestedType.cs#L34 remove this line
for example with the table2 test;
``` CREATE TABLE IF NOT EXISTS test_sub_table_x(id UInt32, params Nested (param_id UInt8, param_val String, sub_nested Nested (param_id UInt8, param_val String))) ENGINE = Memory```

you'll see the parser fails at; https://github.com/DarkWanderer/ClickHouse.Client/blob/master/ClickHouse.Client/Types/TypeConverter.cs#L119

And if you do get past those field name parsing's, the write will fail, since there is an mismatch in the type definition vs. the actual received data.

File: ClickHouse.Client/Formats/BinaryStreamWriter.cs (hunk 4 above)
Comment (olafbuitelaar) [reply ClickHouse.Client/Formats/BinaryStreamWriter.cs]: agreed, it's updated

File: ClickHouse.Client/Formats/BinaryStreamReader.cs (hunk 1 above)
Comment (olafbuitelaar) [reply ClickHouse.Client/Formats/BinaryStreamReader.cs:136]: > `Nested` is an [array of tuples](https://clickhouse.tech/docs/en/sql-reference/data-types/nested-data-structures/nested/), not a single tuple - how does this work when there are multiple values inside?

That is correct, because the Array type is wrapped around the nested type it's still read that way. But an atomic nested value needs to be read as a single tuple.

You can already see this in the unit test; ShouldInsertIntoSubNestedTableViaBulk but i've also extended the unit test; ShouldInsertIntoSubSubNestedTableViaBulk so you can see it works with sub nested values too; https://github.com/olafbuitelaar/ClickHouse.Client/blob/master/ClickHouse.Client.Tests/NestedTableTests.cs#L142

**Comment by DarkWanderer** [comment]
Fixed by #468
